File: EPI/edit_distance.py
def minimum_edit_distance(A, B):
	def E(a, b):
		if a < 0:
			return b + 1
		if b < 0:
			return a + 1
		if cached_e[a][b] == -1:
			if A[a] == B[b]:
				cached_e[a][b] = E(a - 1, b - 1)
			else:
				substitue_last = E(a - 1, b - 1)
				add_last = E(a - 1, b)
				delete_last = E(a, b - 1)
				cached_e[a][b] = 1 + min(substitue_last, add_last, delete_last)
		return cached_e[a][b]
	cached_e = [[-1] * len(B) for _ in range(len(A))]
	return E(len(A) - 1, len(B) - 1)

if __name__ == "__main__":
	A = "b"
	B = "a"
	print(minimum_edit_distance(A,B))

File: IB/dp/edit_distance.cpp
#include <bits/stdc++.h>
using namespace std;
int minimum_edit_distance(string x,string y){
	int n = x.length(), m = y.length();
	vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
	for (int i = 0; i <= n; i++){
		for (int j = 0; j <= m; j++){
			if(i == 0){
				dp[i][j] = j;
			} else if(j == 0){
				dp[i][j] = i;
			}
		}
	}
	for (int i = 1; i <= n; i++){
		for (int j = 1; j <= m; j++){
			if(x[i - 1] == y[j - 1]){
				dp[i][j] = dp[i - 1][j - 1];
			} else {
				dp[i][j] = 1 + min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]);
			}
		}
	}
	return dp[n][m];
}
int main(){

	int t, q;
}

